core/src/main/scala/kafka/log/Log.scala

```diff
@@ -713,9 +713,19 @@ class Log(@volatile var dir: File,
           }
         } else {
           // we are taking the offsets we are given
-          if (!appendInfo.offsetsMonotonic || appendInfo.firstOrLastOffset < nextOffsetMetadata.messageOffset)
+          if (!appendInfo.offsetsMonotonic)
             throw new IllegalArgumentException(s"Out of order offsets found in append to $topicPartition: " +
```

Comment:
Should we be using a different exception here? For example, we should ideally have a test for this and it's easier to write if we can verify it by using the exception type instead of the exception message.

Reply:
You mean we should add a new exception class for this? We have `InvalidOffsetException`, but this one could be thrown from same append() method if we couldn't append to the index, so I think we should not use that one. Maybe we should define `OffsetsOutOfOrderException`?